src/nupic/engine/NetworkFactory.cpp

```diff
@@ -0,0 +1,275 @@
+/* ---------------------------------------------------------------------
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+#include <iostream>
+#include <string>
+
+#include <nupic/engine/Network.hpp>
+#include <nupic/engine/NetworkFactory.hpp>
+#include <nupic/engine/Region.hpp>
+#include <nupic/os/Path.hpp>
+#include <nupic/os/FStream.hpp>
+#include <nupic/utils/Log.hpp>
+#include <nupic/utils/LoggingException.hpp>
+#include <nupic/ntypes/Collection.hpp>
+#include <yaml-cpp/yaml.h>
+
+
+namespace nupic
+{
+  class Input;
+  class Output;
+
+  Network* NetworkFactory::createNetwork(const std::string& path)
+  { 
+    std::string fullPath = Path::normalize(Path::makeAbsolute(path));
+    if (! Path::exists(fullPath)) 
+    {
+      NTA_THROW << "Path " << fullPath << " does not exist";
+    }
+    std::ifstream f(fullPath.c_str());
+    YAML::Parser parser(f);
+    return (createNetworkFromYAML(&parser));
+  }
+
+   
+  Network* NetworkFactory::createNetworkFromYAML(YAML::Parser *parser)
+  { 
+    YAML::Node doc;
+    bool success = parser->GetNextDocument(doc);
+    if (!success) 
+    {
+      NTA_THROW << "Unable to parse YAML document in the specified path.";
+    } 
+
+    if (doc.Type() != YAML::NodeType::Map)
+    {
+      NTA_THROW << "Invalid network structure file -- does not contain a map";
+    }
+      
+    // Should contain Regions and Links only
+    if (doc.size() != 2)
+    {
+      NTA_THROW << "Invalid network structure file -- contains "
+                << doc.size() << " elements when it should contain 2.";
+    }
+
+    // Regions
+    const YAML::Node *regions = doc.FindValue("Regions");
+    const YAML::Node *node;
+    if (regions == nullptr)
+    {
+      NTA_THROW << "Invalid network structure file -- no regions";
+    }
+
+    if (regions->Type() != YAML::NodeType::Sequence) 
+    {
+      NTA_THROW << "Invalid network structure file -- regions element is not a list";
+    }
+
+    auto n = new Network(); // Network to be instantiated by the yaml file.
```

Comment:
Allocate the network on the stack (`Network n;`) to avoid having to delete it manually. You can then make the return type `Network` instead of `Network*` for similar ease of use in the caller. This will get optimized down so there won't actually be a copy of the Network created, it will create it in place on the stack of the calling function which is what you will pretty much always want.